pandas/tests/groupby/test_groupby_dropna.py

```diff
@@ -0,0 +1,236 @@
+import numpy as np
+import pytest
+
+import pandas as pd
+import pandas.util.testing as tm
+
+
+@pytest.mark.parametrize("na_value", [np.nan, None])
+@pytest.mark.parametrize(
+    "dropna, tuples, outputs",
+    [
+        (
+            True,
+            [["A", "B"], ["B", "A"]],
+            {"c": [13.0, 123.23], "d": [13.0, 123.0], "e": [13.0, 1.0]},
+        ),
+        (
+            False,
+            [["A", "B"], ["A", np.nan], ["B", "A"]],
+            {
+                "c": [13.0, 12.3, 123.23],
+                "d": [13.0, 233.0, 123.0],
+                "e": [13.0, 12.0, 1.0],
+            },
+        ),
+    ],
+)
+def test_groupby_dropna_multi_index_dataframe_nan_in_one_group(
+    na_value, dropna, tuples, outputs
+):
+    # GH 3729 this is to test that NA is in one group
+    df_list = [
+        ["A", "B", 12, 12, 12],
+        ["A", na_value, 12.3, 233.0, 12],
+        ["B", "A", 123.23, 123, 1],
+        ["A", "B", 1, 1, 1.0],
+    ]
+    df = pd.DataFrame(df_list, columns=["a", "b", "c", "d", "e"])
+    grouped = df.groupby(["a", "b"], dropna=dropna).sum()
+
+    mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
+    expected = pd.DataFrame(outputs, index=mi)
+
+    tm.assert_frame_equal(grouped, expected, check_index_type=False)
+
+
+@pytest.mark.parametrize(
+    "na_value1, na_value2", [(np.nan, np.nan), (None, None), (np.nan, None)]
+)
+@pytest.mark.parametrize(
+    "dropna, tuples, outputs",
+    [
+        (
+            True,
+            [["A", "B"], ["B", "A"]],
+            {"c": [12.0, 123.23], "d": [12.0, 123.0], "e": [12.0, 1.0]},
+        ),
+        (
+            False,
+            [["A", "B"], ["A", np.nan], ["B", "A"], [np.nan, "B"]],
+            {
+                "c": [12.0, 13.3, 123.23, 1.0],
+                "d": [12.0, 234.0, 123.0, 1.0],
+                "e": [12.0, 13.0, 1.0, 1.0],
+            },
+        ),
+    ],
+)
+def test_groupby_dropna_multi_index_dataframe_nan_in_two_groups(
+    na_value1, na_value2, dropna, tuples, outputs
+):
+    # GH 3729 this is to test that NA in different groups with different representations
+    df_list = [
+        ["A", "B", 12, 12, 12],
+        ["A", na_value1, 12.3, 233.0, 12],
+        ["B", "A", 123.23, 123, 1],
+        [na_value2, "B", 1, 1, 1.0],
+        ["A", na_value2, 1, 1, 1.0],
+    ]
+    df = pd.DataFrame(df_list, columns=["a", "b", "c", "d", "e"])
+    grouped = df.groupby(["a", "b"], dropna=dropna).sum()
+
+    mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
+    expected = pd.DataFrame(outputs, index=mi)
+
+    tm.assert_frame_equal(grouped, expected, check_index_type=False)
```

Comment:
why is check_index-type=False, we almost *never* want to do this, likely something is wrong if you are setting this.

Reply:
emm, this is because of `string` and `np.nan`,  expected will get `string` inferred_type if we use  `mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))` to define `index` while grouped result will provide `mixed` type. 

i feel the behaviour seems correct? maybe i should change how the `index` is defined? @jreback

Reply:
ahh, i think i find the issue, i printed the `dtype` and `inferred_dtype` for both, and they are the same, and then i digged a bit into the `util/testing.py`, and found the issue came from `assert_index_equal`, 
https://github.com/pandas-dev/pandas/blob/c99dfea33612f44e97c2365f78c0ca6d5754a1bc/pandas/_testing.py#L667

it's more like a behaviour issue, now if we do below
```python
tuples = [["A", "B"], ["A", np.nan], ["B", "A"]]
mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
mi.levels
```
Then `mi.levels` will automatically drop `np.nan` in levels, and that\s why the `inferred_type` of expected is `string`, while grouped result is `mixed` which i think is correct. Because if we directly do:
```python
>>> tuples = [["A", "B"], ["A", np.nan], ["B", "A"]]
>>> mi = pd.MultiIndex.from_tuples(tuples, names=list("ab"))
>>> mi.inferred_type
>>> 'mixed'
```
so what is your thoughts on how to fix it? i would consider it a bug in `assert_index_equal`, since grouped result returns the correct type
@jreback @TomAugspurger

Reply:
This seems like an issue with the "grouped" result, no?

Reply:
i think so @WillAyd